### PR TITLE
Fix async pool writer

### PR DESCRIPTION
**Pull request by michalwski**
This PR addresses issue with async writer being too synchronous for reading.

Proposed changes include:
* removed `wait_for_flushing` when a user reads messages from the archive
* renamed `max_packet_size` option to `max_buffer_size`
* in tests use `mam_helper:wait_for_archive_size` in some places instead of `maybe_wait_for_archive`
* documented new behaviour and old config options for async writers

### CODE REVIEW

**Comment by mongoose-im** [comment]
[4882.2](https://travis-ci.org/esl/MongooseIM/jobs/390724692) / Erlang 19.3 / mysql_redis / e0926ddc328f839b9668c2cf1efa48901f7a322d
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1919/4882/mysql_redis.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4882/mysql_redis.19.3/big/ct_run.test%40travis-job-7c94352b-4425-4da2-af74-77ee8cfa7521.2018-06-11_14.20.25/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4882/mysql_redis.19.3/small/ct_run.mongooseim@localhost.2018-06-11_14.13.45/lib.mongooseim.logs/run.2018-06-11_14.13.46/suite.log.html)
<details><summary>login_SUITE:register:end_per_group</summary>


```erlang
{error,
  {{badmatch,
     {error,
       {connection_step_failed,
         {{escalus_session,authenticate},
          {client,<<"admin@localhost">>,escalus_tcp,<0.23828.0>,
            undefined,
            [{username,<<"admin">>},
             {server,<<"localhost">>},
             {host,<<"localhost">>},
             {port,5222},
             {auth,{escalus_auth,auth_plain}},
             {wspath,undefined},
             {username,<<"admin">>},
             {server,<<"localhost">>},
             {password,<<"bruce_almighty">>},
             {stream_id,<<"1F33E08B78F6A40E">>}]},
          [{compression,[<<"zlib">>]},
           {starttls,true},
           {stream_management,true},
           {advanced_message_processing,true},
           {client_state_indication,false},
           {sasl_mechanisms,
             [<<"PLAIN">>,<<"DIGEST-MD5">>,<<"SCRAM-SHA-1">>]},
           {caps,undefined}]},
         {auth_failed,<<"admin">>,
           {xmlel,<<"failure">>,
             [{<<"xmlns">>,
               <<"urn:ietf:params:xml:ns:xmpp-sasl">>}],
             [{xmlel,<<"not-authorized">>,[],[]}]}}}}},
   [{escalus_users,delete_user,2,
      [{file,
         "/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_users.erl"},
       {line,275}]},
    {escalus_users,'-delete_users/2-lc$^0/1-0-',2,
      [{file,
         "/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_users.erl"},
       {line,108}]},
    {test_server...
```

[Report log](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4882/mysql_redis.19.3/big/ct_run.test%40travis-job-7c94352b-4425-4da2-af74-77ee8cfa7521.2018-06-11_14.20.25/big_tests.tests.login_SUITE.logs/run.2018-06-11_14.25.48/login_suite.end_per_group.html)
</details>

**Comment by michalwski** [comment]
Re-opening again since `make dev` was removed on master branch.

**Comment by mongoose-im** [comment]
[4893.2](https://travis-ci.org/esl/MongooseIM/jobs/391172762) / Erlang 19.3 / mysql_redis / c393e7ebab56d75718461b0c95a12989a3110750
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1919/4893/mysql_redis.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4893/mysql_redis.19.3/big/ct_run.test%40travis-job-1101bf43-e51c-4714-b11e-602562f82aee.2018-06-12_09.31.41/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4893/mysql_redis.19.3/small/ct_run.mongooseim@localhost.2018-06-12_09.24.38/lib.mongooseim.logs/run.2018-06-12_09.24.39/suite.log.html)

File: big_tests/tests/mam_SUITE.erl [hunk 1]
```diff
@@ -1203,7 +1203,7 @@ simple_archive_request(Config) ->
         %%   {<<"type">>,<<"chat">>}],
         %%   [{xmlel,<<"body">>,[],[{xmlcdata,<<"OH, HAI!">>}]}]}
         escalus:send(Alice, escalus_stanza:chat_to(Bob, <<"OH, HAI!">>)),
-        maybe_wait_for_archive(Config),
+		mam_helper:wait_for_archive_size(Alice, 1),
```
Comment (arcusfelis) [inline big_tests/tests/mam_SUITE.erl]: indention.

File: big_tests/tests/mam_SUITE.erl [hunk 2]
```diff
@@ -1261,7 +1261,7 @@ simple_text_search_request(Config) ->
         escalus:send(Alice, escalus_stanza:chat_to(Bob, <<"Also my bike broke down so I'm unable ",
                                                           "to return him home">>)),
         escalus:send(Alice, escalus_stanza:chat_to(Bob, <<"Cats are awesome by the way">>)),
-        maybe_wait_for_archive(Config),
```
Comment (arcusfelis) [inline big_tests/tests/mam_SUITE.erl]: we can even remove `maybe_wait_for_archive` at all, if it's not used.
Including `archive_wait` flag.

File: doc/modules/mod_mam.md [hunk 3]
```diff
@@ -79,6 +79,10 @@ These options will only have effect when the `odbc` backend is used:
  When set to `internal`, stores messages and JIDs in internal format.
  **Warning**: Archive MUST be empty to change this option.
 * **async_writer** (boolean, default: `true`) - Enables asynchronous writer that is faster than synchronous but harder to debug.
+  This async writer does not message routing, on the other hand messages are stored in the archive with a delay.
+* **flush_interval** (integer, default: `2000`) How often (in milliseconds) buffered message are flushed to db.
+* **max_buffer_size** (integer, default, `30`) Max size of the buffer for async writer.
```
Comment (arcusfelis) [inline doc/modules/mod_mam.md]: Or just `batch_size`.

File: big_tests/tests/mam_SUITE.erl (hunk 2 above)
Comment (michalwski) [reply big_tests/tests/mam_SUITE.erl]: `maybe_wait_for_archive` is still used. I replaced it with `wait_for_archive_size` in places where it was safe to use (`escalus:fresh_story`) and easy to know how many messages were put to the archive without much digging in the code.
I agree that replacing `maybe_wait_for_archive` with sth more stable and efficient is a good idea. This is better do to in a separate PR though.

File: doc/modules/mod_mam.md (hunk 3 above)
Comment (michalwski) [reply doc/modules/mod_mam.md]: Right, `max_batch_size` is more meaningful. I'll change that.

**Comment by michalwski** [comment]
@arcusfelis I pushed new code with changes related to your comments.

**Comment by mongoose-im** [comment]
[4917.2](https://travis-ci.org/esl/MongooseIM/jobs/392143454) / Erlang 19.3 / mysql_redis / 6ee1cb94ede8d45ab6b56968c8e045742c8b10f3
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1919/4917/mysql_redis.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4917/mysql_redis.19.3/big/ct_run.test%40travis-job-8d1ae7ef-9eb6-4ace-8965-2b338672929e.2018-06-14_10.10.48/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4917/mysql_redis.19.3/small/ct_run.mongooseim@localhost.2018-06-14_10.03.46/lib.mongooseim.logs/run.2018-06-14_10.03.46/suite.log.html)
**OK: 2807** / Failed: 0 / **User-skipped: 222** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[4922.1](https://travis-ci.org/esl/MongooseIM/jobs/392252608) / Erlang 19.3 / internal_mnesia / 41c9fe93af32c91a19e7ebca80414706521fb0f2
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1919/4922/internal_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4922/internal_mnesia.19.3/big/ct_run.test%40travis-job-96c0bcac-f845-4d4b-90fb-3f2e35248c2b.2018-06-14_14.43.08/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1919/4922/internal_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-14_14.36.16/lib.mongooseim.logs/run.2018-06-14_14.36.16/suite.log.html)
**OK: 1068** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Pull request merged**